### PR TITLE
MFU + HFU redux

### PR DESCRIPTION
**Pull request by stas00**
Improving the explanations around MFU, HFU and explaining that it's dangerous to compare these value w/o knowing how they were calculated.

### CODE REVIEW

File: training/performance/README.md [hunk 1]
```diff
@@ -118,8 +131,25 @@ For example [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) published the f
 | 530B  | 56.0% | 57.0% |
 | 1T    | 56.3% | 57.0% |
 
+As you can see, since Megatron-LM was using activation recomputation for these trainings, MFU < HFU.
+
 More recent H100+A100 MFU/HFU numbers have been published [here](https://github.com/mosaicml/llm-foundry/tree/main/scripts/train/benchmarking#mfu-and-hfu).
 
+Now, whenever you see MFU or HFU numbers published you have to be careful comparing those numbers to any other similar numbers, until you know that the same way was used to calculate FLOPS. Since `HFU=Actual_Achieved_FLOPS/Theoretical_FLOPS` and `Theoretical_FLOPS` are fixed. The only variable here is the `
+Actual_Achieved_FLOPS` and since most of the time an estimated value is calculated based on parameter shapes, there are many versions of calculations out there, some of which are slightly to very imprecise. Moreover you don't know how iteration time was measured.
```
Comment (carmocca) [inline training/performance/README.md:139]: One more thing to note is that optimization passes applied by compilers to the training model could also impact the actual achieved value. Often these compilers are not transparent about what they did. Assuming that these passes reduced the number of operations performed reality, the MFU will be optimistic.

On the other hand if the compiler did activation checkpointing silently, the opposite will be true.

TLDR is that you cannot be sure when compilers are in play. What's important is that all of us engineers communicate what we used to measure knowing that it might be flawed.

File: training/performance/README.md [hunk 2]
```diff
@@ -118,8 +131,25 @@ For example [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) published the f
...
+
+So, in order to do a fair comparison the 2 main questions to ask are:
+1. Was the total used floating point operations calculated in the same way?
+2. Was the time component calculated back-to-back of each iteration, including the `DataLoader` and logging, vs only `fwd`+`bwd` parts.
+
+If either one or both of these mismatch then you can't make a fair comparison.
+
+Unfortunately, most of the time papers and blog posts just report the MFU number w/o a link to how it was calculated.
+
+But, do not fear, most of your work will be optimizing your application rather than having a competition of who is faster. And thus as long as you consistently use whatever way you choose to calculate TFLOPS, you will immediately see when your application's performance has improved or degraded, as relative numbers are most important for you.
```
Comment (carmocca) [inline training/performance/README.md:151]: It's not just about competition or marketing. Often you'll hear researchers/engineers say publicly "any code less than X% mfu is bad), without going into all these nuances that you aim to describe here.

So I would reword this more along the lines of:

"Do not fear, if you have trouble comparing your results with competing results keep in mind all the measurement artifacts described above that do not improve the bottom-line throughput. And thus as long as you consistently use ..."

**Review by carmocca — APPROVED** [review]
Thank you so much for this Stas!

File: training/performance/README.md [hunk 3]
```diff
@@ -104,10 +104,23 @@ footnote: For Inference only it'd be: `24Bsh^2 + 4Bs^2h` floating point operatio
 
 Model FLOPS Utilization (MFU) and Hardware FLOPS Utilization (HFU) estimate how well the hardware is being utilized during `forward` and `backward` passes of the model (including any syncing networking overhead and possibly DataLoader IO).
 
-HFU measures the actual FLOPS. For example, the concept [Gradient checkpointing/Activation Recompution](#gradient-checkpointing) repeats all of parts of the `forward` pass a second time, so factually more FLOPS are used. Whereas MFU ignores implementation details and accounts only for the theoretical needs of the computation and thus less accurate.
+```
+MFU = Estimated_Achieved_FLOPS / Theoretical_FLOPS
+HFU = Actual_Achieved_FLOPS / Theoretical_FLOPS
+```
+
+HFU measures the actual FLOPS. For example, the technique of [Gradient checkpointing/Activation Recompution](#gradient-checkpointing) repeats all or some parts of the `forward` pass a second time, so factually more FLOS (FLoating Point Operations) are used. Whereas MFU ignores implementation details and accounts only for the theoretical needs of the computation and thus less accurate.
```
[the remark this answers is not in the document]
Comment (stas00) [reply training/performance/README.md:112]: nice catch, will fix. 

FLOS is what some of us started using to differentiate from FLOPS, because PS usually means Per Second.

File: training/performance/README.md (hunk 1 above)
Comment (stas00) [reply training/performance/README.md]: That's important - thank you, Carlos.

I added your note about compilers. If further improvements can be made please suggest.

File: training/performance/README.md (hunk 2 above)
Comment (carmocca) [reply training/performance/README.md:148]:
```suggestion
But, do not fear, if you have trouble comparing your results with competing results, remember the measurement artifacts described above.
These artifacts do not improve the bottom-line throughput, thus, as long as you consistently use whatever way you choose to calculate TFLOPS, you will immediately see when your application's performance has improved or degraded, as relative numbers are most important for you.
```

Maybe this clarifies what I intended to say

**Comment by stas00** [comment]
Thank you, Carlos. Please let me know if there is anything else to fix, but otherwise we can merge this.

**Comment by carmocca** [comment]
Ship it!
